emberjson/lazy/raw_parser.mojo

```
@@ -0,0 +1,462 @@
+
+    fn __init__(
+        out self,
+        *,
+        ptr: UnsafePointer[Byte, mut=False, origin=origin, **_],
+        length: UInt,
+        idx: UInt = 0,
+    ):
+        # FIXME: temporary rebind
+        self.data = Span[Byte, ptr.origin](
+            ptr=rebind[UnsafePointer[Byte]](ptr), length=length
+        )
+        self.idx = idx
+
+    @always_inline
+    fn __len__(self) -> Int:
+        return len(self.data) - 1 - self.idx
+
+    @always_inline
+    fn has_more(self) -> Bool:
+        return len(self.data) - 1 > self.idx
+
+    @always_inline
+    fn remaining(self) -> String:
+        """Used for debug purposes.
+
+        Returns:
+            A string containing the remaining unprocessed data from parser
+            input.
+        """
+        try:
+            var span = self.data[Int(self.idx) :]
+            var ptr = span.unsafe_ptr()
+            return copy_to_string[True](ptr, ptr + len(span))
+        except:
+            return ""
+
+    @always_inline
+    fn load_chunk(self) -> SIMD8xT:
+        var ptr = self.data.unsafe_ptr()
+        if len(self) < SIMD8_WIDTH:
+            v = SIMD8xT(0)
+            for i in range(len(self)):
+                v[i] = ptr[i]
+            return v
+        return ptr.load[width=SIMD8_WIDTH]()
+
+    @always_inline
+    fn can_load_chunk(self) -> Bool:
+        return len(self) >= SIMD8_WIDTH
+
+    fn parse(mut self, out json: RawJSON[origin]) raises:
+        self.skip_whitespace()
+        if likely(self.has_more()):
+            var b = self.data.unsafe_ptr()[self.idx]
+            if b == `[`:
+                json = self.parse_array()
+                return
+            elif b == `{`:
+                json = self.parse_object()
+                return
+        raise Error("Invalid json")
+
+    fn parse_array(mut self, out arr: RawArray[origin]) raises:
+        self.idx += 1
+        self.skip_whitespace()
+        arr = RawArray[origin]()
+        var ptr = self.data.unsafe_ptr()
```

Why use `Span`, and then extract a ptr from it anyways?

---

Mainly because it carries both the length and the origin information. And we could make it index the span but I think we should avoid the normalization cost (negative indexing) and check it ourselves (because we should raise an error and not only err in `debug_assert` mode) every time to get the maximum performance.

---

The `Parser` itself already carries the origin information, and if we need to dodge index normalization why not just use our own abstraction around `UnsafePointer` where we control the API and implementation details?

---

I don't think there is any reason to build another abstraction other than `Span`, it is literally a pointer and length combo that is everything we need as information. And I don't like reinventing the wheel and having to maintain another abstraction when the stdlib already has the solution.

A bit unrelated but `Span` will also have other helper functions that I'm working on trying to get the stdlib team to accept. I'm actually trying to make it the abstraction on top of `UnsafePointer` that has full vectorization capabilities (vectorized find, vectorized conditional and unconditional apply, vectorized count a boolean function execution, vectorized map, etc.), and make it so that nobody needs to build with `UnsafePointer` when referencing contiguous data.

---

After reading Owen's recent Binary IO proposal I think we should totally stick to `Span`, it will most likely be the common way to express contiguous memory received by a process.

---

A

---

I haven't read it so I'll have to take your word for it

---

It is just a guess though, because every reader or writer will be able to define the buffer type, but I do think that the whole stdlib will converge to using `Span` as the representation for safe (in terms of alloc size) contiguous memory